### PR TITLE
fix(azure-iot-device-amqp): 🐛 multiple AMQP links were being created for telemetry

### LINKED ISSUE #957: Hitting the device side client with a bunch of telemetries immediately will create extra (unused) amqp links
Like the title says.  Using the amqp transport.  Do a bunch of telemetries that don't wait on the callback, one after the other, right after *creating* the device client.  You will get multiple send links.  Only one will get used. Leaving the service to fret over the unused ones.

### LINKED ISSUE #965: warn: MaxListenersExceededWarning: Possible EventEmitter memory leak detected. 11 error listeners added. Use emitter.setMaxListeners() to increase limit
# Context
Server OS
No LSB modules are available.
Distributor ID: Ubuntu
Description:    Ubuntu 18.04.5 LTS
Release:        18.04
Codename:       bionic
- **Docker version:**
19.03

- **Node version:**
node version: v10.23.2
npm version: 6.14.10

- **SDK version details:**
    "azure-iot-common": "1.12.4-lts0720up1",
    "azure-iot-device": "1.17.0-lts0720up1",
    "azure-iot-device-amqp": "1.13.0-lts0720up1",
    "azure-iot-provisioning-device": "1.8.4-lts0720up1",
    "azure-iot-provisioning-device-amqp": "1.8.4-lts0720up1",
    "azure-iot-security-x509": "1.7.4-lts0720up1",
    "azure-storage": "2.10.3",

- **list of installed packages:**
  "dependencies": {
    "@azure/storage-blob": "^12.3.0",
    "async": "2.6.1",
    "azure-iot-common": "1.12.4-lts0720up1",
    "azure-iot-device": "1.17.0-lts0720up1",
    "azure-iot-device-amqp": "1.13.0-lts0720up1",
    "azure-iot-provisioning-device": "1.8.4-lts0720up1",
    "azure-iot-provisioning-device-amqp": "1.8.4-lts0720up1",
    "azure-iot-security-x509": "1.7.4-lts0720up1",
    "azure-storage": "2.10.3",
    "better-queue": "3.8.10",
    "dockerode": "3.2.1",
    "fs-extra": "5.0.0",
    "https-proxy-agent": "2.2.3",
    "lodash": "4.17.19",
    "moment": "2.22.2",
    "mqtt": "2.18.5",
    "nedb": "1.8.0",
    "node-cache": "^5.1.2",
    "systeminformation": "4.30.5",
    "uuid": "3.3.2",
    "winston": "3.2.1"
  },
  "devDependencies": {
    "azure-event-hubs": "^0.2.6",
    "azure-iothub": "^1.9.1",
    "eslint": "^6.8.0",
    "eslint-config-airbnb-base": "^13.1.0",
    "eslint-config-prettier": "^3.3.0",
    "eslint-plugin-import": "^2.14.0",
    "eslint-plugin-prettier": "^3.0.0",
    "jest": "^25.3.0",
    "jest-junit": "^10.0.0",
    "nodemon": "^1.19.1",
    "pem": "^1.14.4",
    "prettier": "^1.15.3",
    "request": "^2.88.0",
    "wtfnode": "^0.6.0",
    "yargs": "^15.3.1"
  },

## Description of the issue
Post device authenticate with IoTHub when the device client either sends telemetry or tries to upload file. We are getting warning on console for 1 time 
warn: MaxListenersExceededWarning: Possible EventEmitter memory leak detected. 11 error listeners added. Use emitter.setMaxListeners() to increase limit
     at _addListener (events.js:256:17)
     at SenderLink.addListener (events.js:272:10)
     at /app/node_modules/azure-iot-device-amqp/dist/amqp.js:376:52
     at /app/node_modules/azure-iot-amqp-base/dist/amqp.js:423:33
     at constructor._onEnter (/app/node_modules/azure-iot-amqp-base/dist/sender_link.js:238:29)
     at constructor.transition (/app/node_modules/machina/lib/machina.js:653:28)
     at constructor.Fsm.(anonymous function) [as transition] (/app/node_modules/machina/lib/machina.js:466:63)
     at constructor.senderOpenEvent (/app/node_modules/azure-iot-amqp-base/dist/sender_link.js:175:36)
     at constructor.handle (/app/node_modules/machina/lib/machina.js:613:25)
     at constructor.Fsm.(anonymous function) [as handle] (/app/node_modules/machina/lib/machina.js:466:63) 

## Code sample exhibiting the issue
Occurring in random installation on few of the nodes, very sporadic in nature

## Console log of the issue

warn: MaxListenersExceededWarning: Possible EventEmitter memory leak detected. 11 error listeners added. Use emitter.setMaxListeners() to increase limit
     at _addListener (events.js:256:17)
     at SenderLink.addListener (events.js:272:10)
     at /app/node_modules/azure-iot-device-amqp/dist/amqp.js:376:52
     at /app/node_modules/azure-iot-amqp-base/dist/amqp.js:423:33
     at constructor._onEnter (/app/node_modules/azure-iot-amqp-base/dist/sender_link.js:238:29)
     at constructor.transition (/app/node_modules/machina/lib/machina.js:653:28)
     at constructor.Fsm.(anonymous function) [as transition] (/app/node_modules/machina/lib/machina.js:466:63)
     at constructor.senderOpenEvent (/app/node_modules/azure-iot-amqp-base/dist/sender_link.js:175:36)
     at constructor.handle (/app/node_modules/machina/lib/machina.js:613:25)
     at constructor.Fsm.(anonymous function) [as handle] (/app/node_modules/machina/lib/machina.js:466:63)

### PR DESCRIPTION
**Pull request by vishnureddy17**
close #965 
close #957

### CODE REVIEW

File: device/transport/amqp/src/amqp.ts [hunk 1]
```diff
@@ -372,20 +376,27 @@ export class Amqp extends EventEmitter implements DeviceTransport {
...
-                  this._d2cLink.on('error', this._d2cErrorListener);
-                  this._d2cLink.send(amqpMessage, handleResult('AMQP Transport: Could not send', sendCallback));
-                }
-              });
-            } else {
+            if (this._d2cLink) {
               debug('using existing d2c link');
               this._d2cLink.send(amqpMessage, handleResult('AMQP Transport: Could not send', sendCallback));
+            } else {
+              debug('waiting for a D2C link');
```
Comment (anthonyvercolano) [inline device/transport/amqp/src/amqp.ts:383]: Tell the story here in a comment.

File: device/transport/amqp/src/amqp.ts [hunk 2]
```diff
@@ -503,6 +514,7 @@ export class Amqp extends EventEmitter implements DeviceTransport {
                 });
               },
               (callback) => {
+                this._amqpLinkEmitter.removeAllListeners('senderLinkAttached');
```
Comment (anthonyvercolano) [inline device/transport/amqp/src/amqp.ts:517]: It's important to be tidy.  thumbs up on seeing this.

File: device/transport/amqp/src/amqp.ts [hunk 3]
```diff
@@ -372,20 +376,27 @@ export class Amqp extends EventEmitter implements DeviceTransport {
...
-                  handleResult('AMQP Transport: Could not send', sendCallback)(err);
-                } else {
-                  debug('got a new D2C link');
-                  this._d2cLink = link;
-                  this._d2cLink.on('error', this._d2cErrorListener);
-                  this._d2cLink.send(amqpMessage, handleResult('AMQP Transport: Could not send', sendCallback));
-                }
-              });
-            } else {
+            if (this._d2cLink) {
```
Comment (anthonyvercolano) [inline device/transport/amqp/src/amqp.ts:379]: You're going to have to add unit tests simply to keep our coverage numbers up AND, this should have some tests anyway.

**Review by anthonyvercolano — APPROVED** [review]
Touch base with me on if you are having an issue creating unit tests.

File: device/transport/amqp/src/amqp.ts [hunk 4]
```diff
@@ -419,9 +434,19 @@ export class Amqp extends EventEmitter implements DeviceTransport {
           disableTwinDesiredPropertiesUpdates: (callback) => this._twinClient.disableTwinDesiredPropertiesUpdates(handleResult('could not disable twin desired properties updates', callback)),
           enableC2D: (callback) => {
             /*Codes_SRS_NODE_DEVICE_AMQP_41_003: [The `enableC2D` method shall attach the C2D link only if it is not already attached.] */
-            if (!this._c2dLink) {
+            if (this._c2dLink) {
+              debug('C2D link already attached, doing nothing....');
+              process.nextTick(callback);
+            } else if (this._c2dLinkAttaching) {
```
Comment (anthonyvercolano) [inline device/transport/amqp/src/amqp.ts:440]: I think you face an issue that in the time betwwen starting to attach the c2d link and it's completion

1) you get a new enable.
2) you go ahead and invoke the callback which gives the implication that c2d is enabled when it's not.
3) the actual enable then finishes 20 ticks later

Honestly, I think your solution to the send line problem was pretty darn good.  Can you use it here?  If you can, you would then be for the most part be repeating code.  Time to put enabling code into a function at that point?  Darn booleans.  Can't trust them.

Comment (vishnureddy17) [reply device/transport/amqp/src/amqp.ts:440]: I see it. Good catch.

File: device/transport/amqp/src/amqp.ts [hunk 5]
```diff
@@ -419,25 +437,37 @@ export class Amqp extends EventEmitter implements DeviceTransport {
...
             /*Codes_SRS_NODE_DEVICE_AMQP_41_003: [The `enableC2D` method shall attach the C2D link only if it is not already attached.] */
-            if (!this._c2dLink) {
-              debug('attaching C2D link');
-              this._amqp.attachReceiverLink(this._c2dEndpoint, null, (err, receiverLink) => {
+            if (this._c2dLink) {
+              debug('C2D link already attached, doing nothing....');
+              process.nextTick(callback);
+            } else {
+              debug('waiting for a C2D link');
+              this._amqpLinkEmitter.once('receiverLinkAttached', (err) => {
```
Comment (anthonyvercolano) [inline device/transport/amqp/src/amqp.ts:445]: You should create a Event Emitter specific for setting up the c2d receiver.  If you share and a c2d enabled at the same time you are setting up the sender, you won't actually set up the c2d receiver.

File: device/transport/amqp/test/_amqp_test.js [hunk 6]
```diff
@@ -1303,14 +1352,21 @@ describe('Amqp', function () {
...
 
         it('calls its callback with an Error if attaching the C2D link fails', function (testCallback) {
-          fakeBaseClient.attachReceiverLink = sinon.stub().callsArgWith(2, new Error('fake failed to attach'));
+          fakeBaseClient.attachReceiverLink = sinon.stub().callsArgWithAsync(2, new Error('fake failed to attach'));
           transport.connect(function () {
             assert(fakeBaseClient.attachReceiverLink.notCalled);
-            transport[testConfig.enableFunc](function (err) {
+            var firstCallbackFired = false
+            function enableCallback (err) {
               assert(fakeBaseClient.attachReceiverLink.calledWith(transport._c2dEndpoint));
```
Comment (anthonyvercolano) [inline device/transport/amqp/test/_amqp_test.js:1360]: Don't you want a called once here?

**Review by anthonyvercolano — APPROVED** [review]

**Comment by anthonyvercolano** [comment]
Looks good.  We'll have to fix the gate though

File: device/transport/amqp/test/_amqp_test.js (hunk 6 above)
Comment (vishnureddy17) [reply device/transport/amqp/test/_amqp_test.js]: Done

**Pull request merged**